### PR TITLE
"Browse By" fixes

### LINKED ISSUE #8646: `webui.itemlist.sort-option.*` configuration properties may be unused
**Describe the bug**
I was asked to add a sort option to the MyDSpace page in 7.4.  Adding a new key in this pattern did nothing.  Eventually I found that adding a value to `config/spring/api/discovery.xml:workspaceConfiguration.searchSortConfiguration.sortFields` *did* work.

There are a number of uses of o.d.sort.SortOption, which reads these properties and builds a Set of SortOption in a static block.  I haven't inspected enough code to know whether these are all dead code, or we have two entirely separate ways of configuring itemlist sort options depending on context.

It's possible that I've thoroughly misunderstood what these properties are for.  But it is curious to have sorting options scattered across two different configuration sources.

**To Reproduce**
1. Make some change to the set of `webui.itemllist.sort-option.*` properties in `dspace.cfg` or `local.cfg`.
2. Restart DSpace
3. See that MyDSpace's sort options are unaltered.

**Expected behavior**
"How to change the set of sorting options on MyDSpace" should be clear, unambiguous, and not confusing.

### LINKED ISSUE #8651: Browse pages don't take into account sort configuration
**Describe the bug**

Browse page results aren't sorted as configured in REST

* The configured sort direction is not taken into account

* The sort type / order cannot be set for metadata indices, even though this should be possible [[according to the documentation](https://wiki.lyrasis.org/display/DSDOC7x/Configuration+Reference#ConfigurationReference-DefiningtheIndexes)](https://wiki.lyrasis.org/display/DSDOC7x/Configuration+Reference#ConfigurationReference-DefiningtheIndexes)

  > If you set a `<sort-type>` to be used, this sort type is not used on the values of the metadata fields but on the order of the items when listing all items that have a specific value of the metadata field.

* The default configuration sorts dates in ascending order, even though the documentation mentions that

  > descending may be useful for date-based indexes

**To Reproduce**

1. Run a local DSpace instance with the following browse configuration
   ```
   webui.browse.index.1 = dateissued:item:dateissued:desc
   webui.browse.index.2 = author:metadata:dc.contributor.*\,dc.creator:text:desc:dateissued
   ```
3. Open the "by issue date" browse page
   * Angular will send a request with `sort=default,ASC`
   * The response will be sorted by ascending date
4. Re-send that request without the `sort` parameter
   * The response will be sorted by descending date, as configured
5. Open the "by author" browse page
   * Angular will send a request with `sort=default,ASC`
   * The REST request will fail with HTTP 500, the following error will appear in the log
     ```
     2023-01-31 13:01:43,944 ERROR 2472f5f8-fc56-476e-9868-25549c647e08 755495a6-7d3f-43a8-9237-e3f25a13bb2b org.dspace.app.rest.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:500)
     java.lang.IllegalStateException: The requested browse doesn't provide direct access to items you must specify a filter
     	at org.dspace.app.rest.repository.BrowseItemLinkRepository.listBrowseItems(BrowseItemLinkRepository.java:82) ~[classes/:?]
       ...
     ```

**Expected behavior**

* When no sort direction is specified, browse pages should sort results as configured in REST.
* When no sort direction is specified, Angular should update the "pagination component" to reflect the default sort direction as configured in REST.

**General thoughts**

While working on #8642 @YanaDePauw noted that browse could be replaced with discovery/Solr facets.
IMO this may be a more worthwhile time investment than maintaining it.

### LINKED ISSUE #8879: `/browses` endpoints return invalid `_links` based on the browseType value.
**Describe the bug**
Discovered while testing #8826.  See [this comment](https://github.com/DSpace/DSpace/pull/8826#pullrequestreview-1442727423).

Currently, all `[dspace.server.url]/api/discover/browses/*` endpoints return invalid `_links` URLs based on the `browseType` value.  
* For instance a `browseType=valueList` doesn't support the `/items` subpath, but still returns that in the `_links` section of the response.
* Similarly, a `browseType=flatBrowse` doesn't support the `/entries` subpath, but still returns that in the `_links` section of the response.

This behavior may seem harmless, but it results in the User Interface making many invalid requests to the REST API anytime you use the Browse by interfaces.  These occur because the UI attempts to follow these invalid "_links" only to get error responses.

* For instance, when using "Browse by Author" (or Browse by Subject) from the UI, I see this 500 error in the DevTools Console:
  ```
  GET http://localhost:8080/server/api/discover/browses/author/items?sort=default,ASC&page=0&size=20&embed=thumbnail 500
  ```
    * The reason is that the "author" browse is a `valueList` and a `valueList` doesn't support `/items`.  So, `/author/items` is invalid, but the UI still tries to use it. This same error occurs for "subject" browse as it's another `valueList`.
* Similarly, when using "Browse by Title" from the UI, I see this 500 error:
  ```
  GET http://localhost:8080/server/api/discover/browses/title/entries?sort=default,ASC&page=0&size=20 500
  ```
    * The reason is that the "title" browse is a `flatBrowse` and a `flatBrowse` doesn't support `/entries`.  So, `/title/entries` is invalid, but the UI still tries to use it.  This same error occurs for "dateissued" browse as it's another `flatBrowse`.

**To Reproduce**
1. Visit http://localhost:4000/.  Open your DevTools and watch the "Network" tab
2. Now, try using any of the Browse By options in the header and notice how the UI will send several extra requests to the REST API which result in 500 errors.  These requests seem to be caused by these invalid links.

**Expected behavior**
No errors should appear in DevTools while using the Browse By features of the UI.

**Related work**
Found during the review of #8826. However, this issue existed on `main` prior to that PR.

### PR DESCRIPTION
**Pull request by buithaihai**
## References
Fix #8651
Fix #8646
Fix #8879

Corresponding frontend PR: DSpace/dspace-angular#2324

## Description

1. For #8651: This PR slightly changes how `BrowseIndex.getDataType()` works, which now accounts for when`webui.browse.index.<X>` uses a sort option that is not of the same data type as the BrowseIndex's own data type by prioritizing the BrowseIndex's. Some other changes also need to be made to follow suit.

    Originally, with config

```
webui.browse.index.2 = author:metadata:dc.contributor.*\,dc.creator:text:desc:dateissued
```
    

  The issue is that we try to sort by "date" type (`dateissued`) while the index itself is of type "text". The "Browse" component does not account for such a case so our changes are needed to ensure Angular call for `/entries` instead of `/items`, which will produce _dspace.log error_ `The requested browse doesn't provide direct access to items you must specify a filter
` when we first access "Browse by Author" via the UI.

2. In addition, we add some comments in `dspace.cfg` to clarify the limitations of `webui.itemlist.sort-option._` for #8646 

3. For #8879: Old code was not updated to use org.dspace.app.rest.utils.Utils#embedOrLinkClassLevelRels as the correct link generator. The update makes sure it is so.

### List of changes in this PR:
1. Updated `BrowseIndex.getDataType()`
2. Changed `org/dspace/browse/BrowseEngine.java:223` so that now it has to use explicit `scope.getBrowseIndex().getSortOption().getType()` if possible:
```java
                value = OrderFormat.makeSortString(value, scope.getFilterValueLang(),
                                                   scope.getBrowseIndex().getSortOption() == null ?
                                                           scope.getBrowseIndex().getDataType() :
                                                           scope.getBrowseIndex().getSortOption().getType());
```
3. Use `bi.getSortOption().getNumber()` if available when we do second level search (check code change for context & details, this is hard to explain properly here).
4. For #8879: Added method RestAddressableModel#isLinkAllowed to check whether a link is valid against an object of type RestAddressableModel. This method is then overridden by org.dspace.app.rest.model.BrowseIndexRest#isLinkAllowed, which validates provided LinkRestagainst its objects. Running this method inUtils#embedOrLinkClassLevelRelsmakes sure we get only valid links for a browse type (_entries_ forvalueListand _items_ forflatBrowse`).

### For Testing:
1. Added the following test case:
- `org.dspace.app.rest.BrowsesResourceControllerIT#testBrowseAuthorByEntriesSortByDateIssued`:  verify that with configuration `webui.browse.index.2 = author:metadata:dc.contributor.*\,dc.creator:text:desc:dateissued`, we can get a list of entries and a list of items sorted by `dateissued` with filtered value "Universe2"
- Modified test `BrowsesResourceControllerIT#findAll` to ensure correct result following the fix for 8879.

2. Testing on a functioning DSpace Installtion:
a. In `dspace.cfg`, change the following config as such:
```
webui.browse.index.2 = author:metadata:dc.contributor.*\,dc.creator:text:desc:dateissued
```
b. On dspace-angular, check Top Navigation bar -> menu **"All of DSpace"** (Browse by) -> **"By Author"**
 - The list of authors should be shown
 - Click on an author should show a list of their items, sorted by `dc.dateissued`, in descending order

c. For 8879: no more 500 error from requests from **[dspace.server.url]/api/discover/browses/*** endpoints

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by buithaihai** [comment]
## Extra note
It seems like we have gone as far as the "Browse" feature permits. To continue improving "Browse" further is like a fool's errand at this point and since a proposal for a replacement with Discovery has been raised, we (DLCorp) would like to express our interest in it and to claim it in the future, should it not be claimed by other parties.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@buithaihai : I tried testing this today, but it is breaking the "Browse by Author" and "Browse by Subject" in the current Angular UI.  So, it's not functional as-is.

Here's how to reproduce the issue:
1. Installed this PR on the backend and deployed.  I've kept my browse configuration to use the default settings in `dspace.cfg`: https://github.com/DSpace/DSpace/blob/main/dspace/config/dspace.cfg#L1140-L1143
3. Deployed latest frontend (`main`) branch, running in production mode (`yarn build:prod; yarn serve:ssr`)
4. Go to "Browse by Subject".  Click on any subject.  The page will not load.  Here's a screenshot of what it looks like:
browse-by-subject-never-loads
5. The same behavior also exists with "Browse by Author".  Click on any author and the page will not load.  (Browse by Title/Date work properly however)

In my browser's DevTools Console, I see this error:
```
ERROR TypeError: Cannot read properties of undefined (reading 'href')
    at main.aa20f532260f4d65.js:1:200299
    at main.aa20f532260f4d65.js:1:1677509
    at i._next (main.aa20f532260f4d65.js:1:1673529)
    at i.next (main.aa20f532260f4d65.js:1:1662321)
    at main.aa20f532260f4d65.js:1:1676238
    at i._next (main.aa20f532260f4d65.js:1:1673529)
    at i.next (main.aa20f532260f4d65.js:1:1662321)
    at main.aa20f532260f4d65.js:1:1677502
    at i._next (main.aa20f532260f4d65.js:1:1673529)
    at i.next (main.aa20f532260f4d65.js:1:1662321)
```

It appears to possibly be referencing this line of code? https://github.com/DSpace/dspace-angular/blob/main/src/app/core/browse/browse.service.ts#L124

Overall, the concepts in this PR seem reasonable.  But, it appears more work is needed to either align the UI code with the changes here... or provide a behavior out of this PR which doesn't break the UI.

**Comment by buithaihai** [comment]
This is part of the fix for #8879, however, it was not working since I removed the `/items` endpoint for `valueList` as suggested here:

> * The reason is that the "author" browse is a `valueList` and a `valueList` doesn't support `/items`

It is actually used, just not the way that is giving 500. It is actually used like this:

`http://192.168.1.179:8080/server/api/discover/browses/author/items?filterValue=<value>`

So the proposed course of action was actually wrong, reinstating the support fixes it. This means that we cannot fix the 500 error for type `valueList`, however.

**NOTE**: In addition, I also remove both `/items` and `/entries` for `hierarchicalBrowse` type since they are producing 500 error on these endpoints, mainly because this type of browse is not configured by `webui.itemlist.sort-option.*` in _dspace.cfg_.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@buithaihai : Re-tested today, and those issues with Browse by Subject & Browse by Author are gone.

However, now I'm seeing this error in my DevTools Console for "Browse by Title":
```
ERROR TypeError: Cannot read properties of undefined (reading 'href')
    at main.ec53954ff112466a.js:1:199687
    at main.ec53954ff112466a.js:1:1679401
    at i._next (main.ec53954ff112466a.js:1:1675421)
    at i.next (main.ec53954ff112466a.js:1:1664219)
    at main.ec53954ff112466a.js:1:1678130
    at i._next (main.ec53954ff112466a.js:1:1675421)
    at i.next (main.ec53954ff112466a.js:1:1664219)
    at main.ec53954ff112466a.js:1:1679394
    at i._next (main.ec53954ff112466a.js:1:1675421)
    at i.next (main.ec53954ff112466a.js:1:1664219)
```

How to reproduce:
1. Install this PR. Run in production mode
2. Open your browse DevTools...keep an eye on the console.
3. Visit "Browse by Title".  The error will appear.

This error doesn't appear to break Browse by Title functionality, but it implies to me that maybe we don't have a good idea yet of how to truly fix #8879.

I'm worried this PR is trying to do too much at this point.  I'd recommend **removing any fixes** related to #8879.  I think those will need more analysis to fix the 500 errors *without* causing a new `TypeError: Cannot read properties of undefined (reading 'href')` in Angular.   I think that analysis may take more time, and isn't likely to be ready for 7.6.

Beyond that, the rest of this PR appears to be working.  It's just that the strategy for fixing #8879 seems to not be correct.

**Comment by buithaihai** [comment]
I think that the fix still works, just that angular code really needs to accommodate the changes. I will be reverting the code for now and see if I can revisit this fix in the future.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@buithaihai  : Sorry, I'm still having issues with getting this PR to work with the Angular UI.  I suspect the reason may be, as you noted, that we may need a corresponding Angular UI PR to fully fix these bugs.

Today, I installed this PR, reindexed (`./dspace index-discovery -b`) and tried to reproduce the bug described in #8651 (by following the "To Reproduce" steps in that PR's description).  **That bug still exists even with this PR installed.**  (NOTE: I tested with the UI running in production mode)
* The Browse by Issue Date in the user interface still sorts (wrongly) by ASC instead of DESC
* The Browse by Author in the user interface still results in an error in the backend logs (dspace.log) and its sort also doesn't appear to be changed?

I do think the code in this PR is a step in the right direction... but I suspect we need to figure out what changes are required in the User Interface before we can move this forward, as it doesn't appear to fix the initially reported bug.

So, I don't see a good reason to merge this for 7.6 right now as it doesn't appear to change the UI behavior.  We either need to find a fix that changes the UI behavior, or consider even replacing the "Browse by" logic with Discovery at some point soon, since these problems seem specific to the old "Browse by" code.  But full replacement is obviously a large project.

Another comment inline below about the fix for #8646.  I think the comments are slightly wrong, unless I am misunderstanding what you mean.

Overall, the direction here seems good... I just don't know how to move it forward until we find a way to get it to work with the User Interface.

**Comment by buithaihai** [comment]
I think delaying this PR for later is fine. I will try to implement the necessary angular changes.

File: dspace/config/dspace.cfg [hunk 1]
```diff
@@ -1174,6 +1174,8 @@ webui.browse.index.4 = subject:metadata:dc.subject.*:text
 # HIDE the option from the sorting controls in the user interface. This can be useful if
 # you need to define a specific date sort for use by the recent items lists,
 # but otherwise don't want users to choose that option.
+# NOTE: These options are now only effective in "Browse by" feature, to modify sort options for other features
+#       (mydspace, workflow, etc) please check out spring/api/discovery.xml
```
[the remark this answers is not in the document]
Comment (buithaihai) [reply dspace/config/dspace.cfg]: I have been trying to explain this in a good way but failed. What I mean is that the equivalent of these `webui.itemlist.sort-option.*` is the `searchSortConfiguration` property that can be found in **discovery.xml**. Therefore if someone wants to change sort options in, like "_My DSpace_", they should go to `searchSortConfiguration` of `bean id="workspaceConfiguration"`.

**Comment by tdonohue** [comment]
Closing as this has not been worked on in over a year and is outdated.  We can reopen later if need be.

**Pull request closed without merging**